File: android/Vastchain_Android_Plugin/face_recognize/src/main/java/ltd/vastchain/face/intercept/InterceptChain.kt
package ltd.vastchain.face.intercept

/**
 * Created by admin on 2021/9/30.
 */
class InterceptChain(
	private val interceptors: List<Interceptor>,
	private val index: Int
) {

	fun proceed(file: String) {
		if (index >= interceptors.size)
			return
		val next = InterceptChain(interceptors, index + 1)
		val interpolator = interceptors[index]
		return interpolator.proceed(file, next)
	}

	fun isLast(): Boolean {
		if (index >= interceptors.size) {
			return true
		}
		return false
	}

	//procced中调用
	fun getNextInterceptor(): Interceptor? {
		if (index >= interceptors.size)
			return null
		return interceptors[index]
	}

	fun currentType(): String {
		var interceptor: Interceptor? = interceptors.firstOrNull { !it.checked() }

		return if (interceptor is VerifyInterceptor) {
			"compare/"
		} else if (interceptor is EyeInterceptor) {
			"eye/"
		} else if (interceptor is MouthInterceptor) {
			"mouth/"
		} else {
			"normal/"
		}
		return "normal/"
	}

	fun isInCompare(): Boolean {
		var interceptor: Interceptor? = interceptors.firstOrNull { !it.checked() }
		return  interceptor is VerifyInterceptor
	}
}

File: ios/VastchainFaceVerify/VastchainFaceVerify/Classes/InterceptChain.m
//
//  InterceptChain.m
//  VastchainPlugin_Example
//
//  Created by cxd on 2021/10/13.
//  Copyright © 2021 chenxiandiao. All rights reserved.
//

#import "InterceptChain.h"
#import "VerifyInterceptor.h"
#import "EyeInterceptor.h"
#import "MouthInterceptor.h"

@implementation InterceptChain

- (id) init:(NSArray *)intercepotrs index: (NSInteger)index {
    self = [super init];
    self.interceptors = intercepotrs;
    self.index = index;
    return self;
}

- (void)procced:(NSString *)file {
    if (self.index >= self.interceptors.count) {
        return;
    }
    NSLog(@"procced");
    InterceptChain *next = [[InterceptChain alloc]init:self.interceptors index:self.index+1];
    Interceptor* interceptor = self.interceptors[self.index];
    [interceptor procceed:file chain:next];
}

- (NSString *)currentType {
    for (Interceptor *interceptor in _interceptors) {
        if (![interceptor checked]) {
            if ([interceptor isKindOfClass:[VerifyInterceptor class]]) {
                return @"compare";
            } else if ([interceptor isKindOfClass:[EyeInterceptor class]]) {
                return @"eye";
            } else if ([interceptor isKindOfClass:[MouthInterceptor class]]) {
                return @"mouth";
            }
            return @"compare";
        }
    }
    return @"compare";
}


- (BOOL)isLast {
    if (self.index>=self.interceptors.count) {
        return YES;
    }
    return NO;
}

- (Interceptor *)getNextInterceptor {
    if (self.index>=self.interceptors.count) {
        return nil;
    }
    return self.interceptors[self.index];
}

@end
